source/jormungandr/jormungandr/scenarios/experimental.py

```diff
@@ -244,90 +400,89 @@ def call_kraken(self, request_type, request, instance, krakens_call):
-                g.origins_fallback[dep_mode] = self._get_stop_points(instance,
-                                                                     g.requested_origin, dep_mode,
-                                                                     get_max_fallback_duration(request, dep_mode),
-                                                                     request)
-
-                #Fetch all the stop points of this stop_area and replaces all the durations by 0 in the table
-                #g.origins_fallback[dep_mode]
-                _update_crowfly_duration(instance, g.origins_fallback, dep_mode, request['origin'],
-                                         crow_fly_stop_points)
-
-            if arr_mode not in g.destinations_fallback and request.get('max_duration', 0):
-               g.destinations_fallback[arr_mode] = self._get_stop_points(instance,
-                                                                         g.requested_destination, arr_mode,
-                                                                         get_max_fallback_duration(request, arr_mode),
-                                                                         request,
-                                                                         reverse=True)
-
-               #Fetch all the stop points of this stop_area and replaces all the durations by 0 in the table
-               #g.destinations_fallback[arr_mode]
-               _update_crowfly_duration(instance, g.destinations_fallback, arr_mode, request['destination'],
-                                        crow_fly_stop_points)
+        worker = AsyncWorker(instance, krakens_call, request)
+
+        if request.get('max_duration', 0):
+            orig_futures, dest_futures = worker.get_crowfly_futures(g.requested_origin, g.requested_destination)
+            gevent.joinall(orig_futures + dest_futures)
+            for future in gevent.iwait(orig_futures):
+                g.origins_places_crowfly.update(future.get())
+            for future in gevent.iwait(dest_futures):
+                g.destinations_places_crowfly.update(future.get())
+
+            orig_futures, dest_futures = worker.get_routing_matrix_futures(g.requested_origin,
+                                                                           g.requested_destination,
+                                                                           g.origins_places_crowfly,
+                                                                           g.destinations_places_crowfly)
+            gevent.joinall(orig_futures + dest_futures)
+            for future in gevent.iwait(orig_futures):
+                g.origins_fallback.update(future.get())
+            for future in gevent.iwait(dest_futures):
+                g.destinations_fallback.update(future.get())
+
+            orig_futures, dest_futures = worker.get_update_crowfly_duration_futures()
```

Comment:
Hmm, this functions update the sn_routing_matrix if the stop_points is contained by stop_area, so we can only do this after the computation of sn_routing_matrix